Extending wpa supplicant mock with paf support by introducing nan simulator

filipwasil opened the pull request:
#### Summary

Added wifipaf commissioning feature to WpaSupplicantMock to be able to commission without real WIFI hardware.

- NANSimulator
- Option in run_test_suite.py
- Test entry in tests.yaml
- Connectivity manager now tracks mPafChannelAvailable properly (it was blocking commissioning)
- HandleNonConcurrentConnectNetwork not called when no CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
- dbus callback signature changed to correct one

I was basing on already existing BLE-WiFi approach.

#### Related issues

https://github.com/project-chip/connectedhomeip/issues/41577

There is a ticket here, but no further information. I was contacting few people in regard to it but did not get any specific info about any ongoing efforts in this regard.

#### Testing

Build
```
source scripts/activate.sh && scripts/build/build_examples.py     --target linux-x64-all-clusters     --target linux-x64-chip-tool     build
```

Test (for ble-wifi and wifi-paf commissioning)
```
scripts/tests/run_test_suite.py --runner chip_tool_python --chip-tool out/linux-x64-chip-tool/chip-tool --target TestOperationalState --log-level debug run --all-clusters-app out/linux-x64-all-clusters/chip-all-clusters-app --commissioning wifi-paf

scripts/tests/run_test_suite.py --runner chip_tool_python --chip-tool out/linux-x64-chip-tool/chip-tool --target TestOperationalState --log-level debug run --all-clusters-app out/linux-x64-all-clusters/chip-all-clusters-app --commissioning ble-wifi
```

Test result:
```
...
DEBUG   [CTL] Commissioning complete for node ID 0x0000000012344321: success
DEBUG   [TOO] Device commissioning completed with success
...
```

```
``` 

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

arkq reviewed (dismissed):
This PR should wait for slight refactoring of the `linux.py`. As pointed out in https://github.com/project-chip/connectedhomeip/pull/42473#pullrequestreview-3631568469 the linux.py file becomes too big with "utils" like vibe. It should be split into logical components under "linux" Python submodule and placed in the `src/python_testing/matter_testing_infrastructure/matter/testing` so it would be possible to use mocks in Python tests as well.

filipwasil committed:
commissioning and connectivity improvements related to wifi-paf

filipwasil commented:
> This PR should wait for slight refactoring of the `linux.py`. As pointed out in [#42473 (review)](https://github.com/project-chip/connectedhomeip/pull/42473#pullrequestreview-3631568469) the linux.py file becomes too big with "utils" like vibe. It should be split into logical components under "linux" Python submodule and placed in the `src/python_testing/matter_testing_infrastructure/matter/testing` so it would be possible to use mocks in Python tests as well.

Done

arkq commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/wifi.py:
```diff
@@ -111,13 +275,245 @@ async def Disconnect(self) -> None:
         @State.setter_private
         def State_setter(self, value: str) -> None:
             self.state = value
+            self.PropertiesChanged.emit({'State': ('s', value)})
```
Is this custom `PropertiesChanged` signal required? wpa_supplicant has a generic `org.freedesktop.DBus.Properties.PropertiesChanged` interface as well and as far as I know signals are delivered twice, once on the custom `PropertiesChanged` signal and once on generic `org.freedesktop.DBus.Properties`. If possible we should use generic interface instead of legacy one.

filipwasil committed:
brought back original signal name

filipwasil committed:
removed unused emits from NaN simulator

filipwasil committed:
removed the c-like size store

restyled-commits committed:
Restyled by isort

filipwasil commented on scripts/tests/run_test_suite.py:
```diff
@@ -418,6 +424,12 @@ def build_app(arg_value: Path | None, kind: SubprocessKind, key: str) -> Subproc
+        raise click.BadOptionUsage("wifi-paf", "Option --wifi-paf is available on Linux platform only")
+
+    if ble_wifi and wifi_paf:
+        raise click.BadOptionUsage("wifi-paf", "Options --ble-wifi and --wifi-paf are mutually exclusive")
```
[EDIT] its "wifipaf-wifi" now

filipwasil committed:
fixes in type hints

filipwasil commented on src/platform/Linux/ConnectivityManagerImpl_NetworkManagementWpaSupplicant.cpp:
```diff
@@ -368,6 +368,10 @@ void ConnectivityManagerImpl::_OnWpaPropertiesChanged(WpaSupplicant1Interface *
         TEMPORARY_RETURN_IGNORED DeviceLayer::SystemLayer().ScheduleLambda([]() { ConnectivityMgrImpl().UpdateNetworkStatus(); });
         NotifyWiFiConnectivityChange(kConnectivity_Lost);
 
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF
```
If WPASupplicant state is "disconnected", then the mPafChannelAvailable = true
If WPASupplicant state is "completed", then  the mPafChannelAvailable = true

Succes route:
1. _ConnectWiFiNetworkAsync
When wpa_supplicant_1_interface_call_add_network_sync was called and succeeded, The mPafChannelAvailable is kept false. 
2. _OnWpaInterfaceScanDone
Is the only place when it can be set mPafChannelAvailable to true.

Failed route:
1. _ConnectWiFiNetworkAsync
When wpa_supplicant_1_interface_call_add_network_sync was called and failed, The mPafChannelAvailable is set to true instantly. 

New route:
1. In _OnWpaPropertiesChanged either wpa supplicant **state** is **disconnected** or **completed** will set mPafChannelAvailable  to true.

Justification for this change:

When wpa_supplicant_1_interface_call_add_network_sync is called and succeeds, but no _ConnectWiFiNetworkAsync is called afterward, the mPafChannelAvailable will remain false regardless what is happening with WpaSupplicant state.

filipwasil committed:
cleanup in sdk files. added BSS list

filipwasil committed:
Too much changes in one PR

restyled-commits committed:
Restyled by autopep8

filipwasil committed:
Removed not used signals

filipwasil committed:
Improved quality of python code

filipwasil committed:
refactor

filipwasil committed:
do not store reference to mock in wpa supplicant mock

arkq commented on src/python_testing/matter_testing_infrastructure/matter/testing/linux/wifi.py:
```diff
@@ -21,77 +21,249 @@
+import logging
 import threading
-from typing import Union
+from typing import Any, Dict, Optional, Tuple, Union
```
Use `dict` and `tuple` for typeinfo. Instead of `Optional` use new syntax `| None`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=42481) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=42481) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: filipwasil<br/>:x: Filip Wasil<br/><hr/>**Filip Wasil** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=42481) it.</sub>

filipwasil committed:
Updates after review:

- Tuple -> tuple
- Dict -> dict
- No timeouts
- interfaces_info -> interfaces_names

filipwasil committed:
Better python, removed not needed GVariant checks

filipwasil committed:
fix: scan done emited from scan method

filipwasil committed:
fix: moved network creation to interface to avoid indices

filipwasil committed:
Simplifications

filipwasil committed:
further simplifications

filipwasil committed:
Introducing Scanning property to WpaSupplicantMock

filipwasil committed:
Extra debug info when started scanning

restyled-commits committed:
Restyled by autopep8

filipwasil committed:
fixing mypy issues

filipwasil committed:
mypy fixes - constructors

restyled-commits committed:
Restyled by autopep8

filipwasil committed:
reverted constructor changes

filipwasil committed:
Update src/python_testing/matter_testing_infrastructure/matter/testing/linux/wifi.py

Co-authored-by: Arkadiusz Bokowy <arkadiusz.bokowy@gmail.com>

Alami-Amine commented on .github/workflows/tests.yaml:
```diff
@@ -466,6 +466,28 @@ jobs:
+                     --iterations 1 \
+                     --test-timeout-seconds 120 \
+                     --tool-path chip-tool:./objdir-clone/linux-x64-chip-tool${CHIP_TOOL_VARIANT}-${BUILD_VARIANT}/chip-tool \
+                     --app-path all-clusters:./objdir-clone/linux-x64-all-clusters-${BUILD_VARIANT}/chip-all-clusters-app \
```
Could we test against ASAN built All-Clusters? we are migrating all apps in tests.yaml to asan-clang very soon
```suggestion
                     --app-path all-clusters:./objdir-clone/linux-x64-all-clusters-${BUILD_VARIANT}-asan-clang/chip-all-clusters-app \
```

arkq replied:
Isn't it going to be picked up automatically by `${BUILD_VARIANT}`?

Anyway, it's not possible to make that change without changing all other places (builds and other test jobs).
